enhancements/api-review/gcp_user_defined_labels.md

```diff
@@ -0,0 +1,610 @@
+  by label in cloud billing.
+- As an OpenShift administrator, I want to have tags added to all resources created
+  in GCP by OpenShift, so that I can use it for creating IAM policies or to configure 
+  organization policy constraints or to refine resource usage and cost data.
+
+### Goals
+
+- The administrator or service (in the case of Managed OpenShift) installing OpenShift
+  can configure a list of up to 32 user-defined labels in the OpenShift installer generated
+  install config, which is referenced and applied by installer and the in-cluster operators
+  on the GCP resources during creation.
+- Labels must be applied at creation time, in an atomic operation. Labels will not be added post cluster creation.
+- The administrator or service (in the case of Managed OpenShift) installing OpenShift can configure up to 50 tags
+  in the install config. The tags are added by installer and the in-cluster operators while creating GCP resources.
+
+### Non-Goals
+
+- Support for update or delete operation on labels or tags after creation of GCP resource.
+
+## Proposal
+
+***GCP Labels*** 
+
+A label of the form `kubernetes-io-cluster-<cluster_id>:owned` will be added to every
+resource created by OpenShift to enable administrator to differentiate the resources
+created for OpenShift cluster. An administrator is not allowed to add or modify the label
+having the prefix `kubernetes-io` or `openshift-io` in the key. The same can be found
+applied to other cloud platform resources which supports labels/tags for ex: AWS, Azure.
+
+New `userLabels` field will be added to `platform.gcp` of install-config for the user
+to define the labels to be added to the resources created by installer and in-cluster operators.
+
+If `platform.gcp.userLabels` in the install-config has any labels defined, then these labels
+will be added to all the GCP resources created by OpenShift, provided all the defined
+labels meet all the below conditions or else the cluster creation will fail.
+1. A label key and value must have minimum of 1 character and can have maximum of 63 characters.
+2. A label key and value must contain only lowercase letters, numeric characters,
+   underscores, and dashes.
+3. A label key must start with a lowercase letter.
+4. Each resource can have multiple labels, up to a maximum of 32.
+   - GCP supports a maximum of 64 labels per resource. Restricting the number of
+    user defined labels to 32 and reserving 32 for Openshift's internal use.
+
+All in-cluster operators that create GCP resources (Cluster Infrastructure ,Storage) will
+apply these labels during resource creation.
+
+The userLabels field is intended to be set at install time and is considered immutable.
+Components that respect this field must only ever add labels that they retrieve from this
+field to cloud resources, they must never remove labels from the existing underlying cloud
+resource even if the labels are removed from this field(despite it being immutable).
```

Comment:
The Filestore driver adds a `storage_gke_io_created-by: filestore_csi_storage_gke_io` label automatically. I'm not sure we would be able to change this behaviour since it's already in the driver and users may depend on it.